File: 22-generate-parentheses/22-generate-parentheses.cs
public class Solution {
  List<string> result = new();
  private void BackTrack(string s, int left, int right) {
    if (left == 0 && right == 0) {
      result.Add(s);
    }
    if (left > 0)
      BackTrack(s+'(', left - 1, right);
    if (right > left)
      BackTrack(s+')', left, right - 1);
  }

  public IList<string> GenerateParenthesis(int n) {
    BackTrack("", n, n);
    return result;
  }
}

File: 22-generate-parentheses/22-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
  const result = []
  const backTrack = (s, l, r) => {
    if (l === 0 && r === 0) {
      result.push(s)
    }
    if (l > 0) backTrack(s + '(', l-1, r)
    if (l < r) backTrack(s + ')', l, r-1)
  }
  backTrack('', n, n);
  return result;
};